src/Entity/FeedsMigrateImporter.php

```diff
@@ -29,165 +33,215 @@
+  public function getOriginalMigration() {
+    if (!isset($this->originalMigration)) {
+      $this->originalMigration = Migration::load($this->migrationId);
+    }
+
+    return $this->originalMigration;
+  }
 
-    foreach ($fetcher_plugins->getDefinitions() as $definition) {
-      if ($definition['parent'] == $source_configuration['data_fetcher_plugin']) {
-        /** @var \Drupal\feeds_migrate\DataFetcherFormInterface $fetcher_instance */
-        $fetcher_instance = $fetcher_plugins->createInstance($definition['id']);
-        $fetcher_instance->alterMigration($this, $this->migration);
-      }
+  /**
+   * {@inheritdoc}
+   */
+  public function getMigration() {
+    if (!isset($this->migration)) {
+      /* @var \Drupal\migrate_plus\Entity\MigrationInterface $altered_migration */
+      $altered_migration = $this->migration = clone $this->getOriginalMigration();
+
+      $source = array_merge($this->originalMigration->get('source'), $this->migrationConfig['source'] ?? []);
+      $altered_migration->set('source', $source);
+      $destination = array_merge($this->originalMigration->get('destination'), $this->migrationConfig['destination'] ?? []);
+      $altered_migration->set('destination', $destination);
     }
+
+    return $this->migration;
   }
 
   /**
-   * Alter the authentication method from the configured plugin.
+   * {@inheritdoc}
    */
-  protected function alterAuthentication() {
-    $source_configuration = $this->migration->getSourceConfiguration();
+  public function calculateDependencies() {
+    $dependencies = parent::calculateDependencies();
+    // TODO add dependency on migration entity.
+    return $dependencies;
+  }
 
-    if (empty($source_configuration['authentication']['plugin'])) {
-      return;
+  /**
+   * {@inheritdoc}
+   */
+  public function needsImport() {
+    $request_time = Drupal::time()->getRequestTime();
+    if ($this->importFrequency != FeedsMigrateImporterInterface::SCHEDULE_NEVER && ($this->getLastRun() + $this->importFrequency) <= $request_time) {
+      return TRUE;
     }
 
-    $auth_plugins = \Drupal::service('plugin.manager.feeds_migrate.authentication_form');
-    foreach ($auth_plugins->getDefinitions() as $definition) {
-      if ($definition['parent'] == $source_configuration['authentication']['plugin']) {
-        /** @var \Drupal\feeds_migrate\AuthenticationFormInterface $auth_instance */
-        $auth_instance = $auth_plugins->createInstance($definition['id']);
-        $auth_instance->alterMigration($this, $this->migration);
-      }
-    }
+    return FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getBatchExecutable() {
+    $messenger = new MigrateMessage();
+    return new FeedsMigrateBatchExecutable($this, $messenger);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getExecutable() {
+    $messenger = new MigrateMessage();
+    return new FeedsMigrateExecutable($this, $messenger);
   }
```

Comment:
Does this second executable import everything in one go? For very large imports, would it be viable to divide the import into chunks, to avoid PHP timeouts? Or isn't that needed for Migrate? I looked briefly at MigrateExecutable this week and noticed a comment about avoiding PHP timeouts. I haven't inspected it closely enough yet to understand how that works.

Reply:
We're using the batch executable for importing and the normal one for rolling back. PHP timeouts are definitely something we should keep in mind moving forward